source/common/config/delta_subscription_state.h

```diff
@@ -22,11 +22,61 @@ namespace Config {
 // blissfully unaware of each other's existence, even when their messages are
 // being multiplexed together by ADS.
+//
+// There are two scenarios which affect how DeltaSubscriptionState manages the resources. First
+// scenario is when we are subscribed to a wildcard resource, and other scenario is when we are not.
+//
+// Delta subscription state also divides the resources it cached into three categories: requested,
+// wildcard and ambiguous.
+//
+// The "requested" category is for resources that we have explicitly asked for (either through the
+// initial set of resources or through the on-demand mechanism). Resources in this category are in
+// one of two states: "complete" and "waiting for server".
+//
+// "Complete" resources are resources about which the server sent us the information we need (for
+// now - just resource version).
+//
+// The "waiting for server" state is either for resources that we have just requested, but we still
+// didn't receive any version information from the server, or for the "complete" resources that,
+// according to the server, are gone, but we are still interested in them - in such case we strip
+// the information from the resource.
+//
+// The "wildcard" category is for resources that we are not explicitly interested in, but we are
+// indirectly interested through the subscription to the wildcard resource.
+//
+// The "ambiguous" category is for resources that we stopped being interested in, but we may still
+// be interested indirectly through the wildcard subscription - resources in these category are
+// "waiting" for the config server to confirm their status.
+//
+// Please refer to drawings (non-wildcard-resource-state-machine.png and
+// (wildcard-resource-state-machine.png) for visual depictions of the resource state machine.
+//
+// In the "no wildcard subscription" scenario all the cached resources should be in the "requested"
+// category. Resources are added to the category upon the explicit request and dropped when we
+// explicitly unsubscribe from it. Transitions between "complete" and "waiting for server" happen
+// when we receive messages from the server - if a resource in the message is in "added resources"
+// list (thus contains version information), the resource becomes "complete". If the resource in the
+// message is in "removed resources" list, it changes into the "waiting for server" state. If a
+// server sends us a resource that we didn't request, it's going to end up in the "wildcard"
+// category. Such resources are dropped when the server send us a message with the resource in
+// "removed resources" list. But this normally should not happen.
+//
+// In the "wildcard subscription" scenario, "requested" category is the same as in "no wildcard
+// subscription" scenario, with one exception - the unsubscribed "complete" resource is not removed
+// from the cache, but it's moved to the "ambiguous" resources instead. At this point we are waiting
```

Comment:
This seems a little sub-optimal, because it means that if a client is subscribed to the wildcard and to another explicit resource, if it unsubscribes from that explicit resource, it cannot remove the resource from its cache unless the server explicitly sends a response with the resource in the `removed_resources` list.  This could lead to the cache growing without bound.

Maybe we should add something to the xDS spec that requires the delta server to send the resource name in `removed_resources` in response to a client unsubscription, but only if the client is subscribed to the wildcard.

@htuch, WDYT?

Reply:
> This seems a little sub-optimal, because it means that if a client is subscribed to the wildcard and to another explicit resource, if it unsubscribes from that explicit resource, it cannot remove the resource from its cache unless the server explicitly sends a response with the resource in the `removed_resources` list. This could lead to the cache growing without bound.
> 
> Maybe we should add something to the xDS spec that requires the delta server to send the resource name in `removed_resources` in response to a client unsubscription, but only if the client is subscribed to the wildcard.
> 
> @htuch, WDYT?

Yeah, my assumption was that in such case (we are subscribed to wildcard and just unsubscribed from some explicit resource) the server will send us a response with the explicit resource in `added_resources` if the explicit resource is also a part of the wildcard set (so we would move the resource from ambiguous to wildcard) or otherwise in `removed_resources` if the explicit resource is not a part of the wildcard set (so we would drop the resource from the cache).

But yeah, probably it's best to add such a wording to xDS spec.

Reply:
> the server will send us a response with the explicit resource in added_resources if the explicit resource is also a part of the wildcard set (so we would move the resource from ambiguous to wildcard) 

As mentioned above -- why cache a resource if we receive a complete copy from the server anyway?

> But yeah, probably it's best to add such a wording to xDS spec.

I don't think this is current behaviour and it needs to be documented.

Reply:
Good point.  It seems inefficient to require the server to explicitly resend the resource if it *is* included in the wildcard, so I wasn't thinking it would do that.  But I guess the client would need some signal to tell it to remove the resource from the "ambiguous" list in this case.

I've put together #17983 to update the spec.

Reply:
@dmitri-d In the incremental protocol variant, the server does not need to resend any individual resource unless that resource has changed, so we won't necessarily receive a complete copy from the server anyway.

Reply:
Personally I'm not a fan of maintaining a state for a deleted resource -- we are now mandating an implementation for other xds clients. This also creates an ambiguity between unsubscribing from a resource and resource being unsubscribed from b/c of expired ttl: for the former we keep the state around, for the latter we remove the state (the unsubscribe request looks the same though). While its certainly possible to distinguish between the two, it would make server implementation more complicated. Not sure this complexity is worth it for what I think is a rather weird corner-case.

Reply:
> Just realised: This change might break current server implementations b/c of difference in handling of ttl expirations vs. unsubscribes.

Could you expand on it, please? Right now I'm not sure if and how to address this. Thanks.

Reply:
I'm not sure what the problem is with TTLs here.  Can you say more about that?

With regard to the broader issue, I agree that the presence of "ambiguous" state is fairly ugly.  But I think the root cause of the ugliness is actually the wildcard subscription semantics, which were not designed to coexist with per-resource subscriptions and do not provide any way for the client to know which returned resource is actually associated with the wildcard subscription.  Given that inherent wire-level protocol problem, I don't see any choice other than to mandate a certain behavior for clients.

I will also note two things that I think make this more tolerable.  First, the new xdstp: naming scheme replaces wildcard subscriptions with collection resources, and collection resources do not have the above problem: the wire protocol clearly indicates which returned resources are associated with which collection, so there is no ambiguity.  This means that this problem will go away as we migrate to the new naming scheme.

Second, even for legacy clients and servers, this problem comes up only in a case where the client is subscribing to both the wildcard and to other resources on the same stream.  In general, that will occur only when something like on-demand CDS is used, which means that clients that don't implement on-demand CDS will not need to worry about this.

Reply:
Currently, when a resource ttl expires, the resource gets deleted. It happens regardless of whether the resource is a part of wildcard subscription or not. IIRC the original conversation, the assumption was that the server will just resend a resource if the server missed an update.

With the changes in this PR, the client keeps the state for resources some of the time now. There isn't a way for the server to distinguish between unsubscribe request due to a ttl expiration from a regular unsubscribe based on data present in discovery request. As we expect different responses from the server (a full resource in case of ttl expiration, a resource name otherwise),   the server is now required to distinguish between the two based on its internal state. which wasn't a requirement before. b/c of the above the changes in this PR would require a corresponding update in server implementations.

Reply:
I don't think the client actually generates an unsubscribe request when a TTL expires.  A TTL is basically a server's way of saying "consider the resource to be removed at a particular time without my having to explicitly tell you that it's been removed".  But a server indicating that a resource has been removed does that alter the fact that a client is subscribing to it.  In effect, you can think of "does not exist" as just another possible value for a given resource; while a client is subscribed to a resource, the server is obligated to send updates whenever the value of that resource changes, so "resource previously did exist and now does not" and "resource previously did not exist and now does" are both just special cases of the resource's value changing.

A server that sends a resource with a TTL knows that the client will consider the resource to not exist when the TTL expires.  If the server wants the client to continue using the resource after that, it needs to resend the resource with a new TTL value before the original TTL expires.  But this should not be hard for the server to figure out, because the server knows that it sent the resource with a TTL in the first place.  And I think this behavior is required independently of the change we're talking about here.

In the change we're talking about here, all we're saying is that a server that supports both wildcard and non-wildcard subscriptions on the same stream needs to know that if the client unsubscribes to a non-wildcard resource, the server needs to explicitly indicate whether that resource is still relevant to the client because of the wildcard or whether the client can remove the resource.  I think this change is completely independent of TTL behavior.

Reply:
> I don't think the client actually generates an unsubscribe request when a TTL expires.

Duh, indeed there's no unsubscribe request on ttl expiration. Ignore what I said.